test/e2e/scheduling/predicates.go

```diff
@@ -724,6 +727,231 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
 		ginkgo.By(fmt.Sprintf("Trying to create another pod(pod5) with hostport %v but hostIP %s on the node which pod4 resides and expect not scheduled", port, hostIP))
 		createHostPortPodOnNode(ctx, f, "pod5", ns, hostIP, port, v1.ProtocolTCP, nodeSelector, false)
 	})
+	ginkgo.Context("PodAffinity/PodAntiAffinity Filtering", func() {
+		ginkgo.It("PodAffinity: Pods from the same replicaset are taken into account by matchLabelKeys during the deployment's rolling upgrade [Feature:MatchLabelKeysInPodAffinity] [alpha]", func(ctx context.Context) {
+			// It's a real-world usecase of matchLabelKeys.
+			// Create a deployment with 3 replicas, run rolling upgrade, and see new replicas are not affected by the old replicas.
+
+			e2eskipper.SkipUnlessNodeCountIsAtLeast(2)
+			ginkgo.By("make the nodes have balanced resource usage ratio")
+			err := createBalancedPodForNodes(ctx, f, cs, ns, nodeList.Items, podRequestedResource, 0.5)
+			framework.ExpectNoError(err)
+
+			k := v1.LabelHostname
+			// For the first deployment:
+			// all Pods should be scheduled on the same Node.
+			//
+			// For the second deployment:
+			// When the first Pod from the second deployment is scheduled, no Pods in the cluster matches with the hard pod affinity due to the matchLabelKeys.
+			// So, the inter-pod affinity filter plugin allows the first Pod to go anywhere.
+			// And, the inter-pod affinity score plugin prefer the Node which the first deployment's Pod do NOT exist because of the soft antiaffinity.
+			//
+			// As a result, all Pods from the second deployment should be scheduled on the same Node but the one different from which the first deployment's Pod exist.
+			affinity := &v1.Affinity{
+				PodAffinity: &v1.PodAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+						// affinity; should be scheduled on the Node which has security:S1 Pod.
+						{
+							LabelSelector: &metav1.LabelSelector{
+								MatchExpressions: []metav1.LabelSelectorRequirement{
+									{
+										Key:      "security",
+										Operator: metav1.LabelSelectorOpIn,
+										Values:   []string{"S1"},
+									},
+								},
+							},
+							TopologyKey:    k,
+							Namespaces:     []string{ns},
+							MatchLabelKeys: []string{"pod-template-hash"},
+						},
+					},
+				},
+				PodAntiAffinity: &v1.PodAntiAffinity{
```

Comment:
There are other ideas. But, I'm thinking use preferred PodAntiAffinity is the most simple way among them.

- Use taint/toleration: add the toleration only on the old replicaset, and add the taint on Node which all old replicaset Pods are scheduled on. New replicaset Pods cannot get scheduled on that Node due to the taint. So, if matchLabelKeys is working correctly, the new replicaset Pods will be scheduled on different Node; otherwise they'll be unschedulable. 
- Use PodTopologySpread: add the topologySpread which forces all Pods to go to different Nodes. If matchLabelKeys is working correctly, the new replicaset Pods will be scheduled on different Node; otherwise they'll be unschedulable. 

What do you think? 💭